### PR TITLE
Re-Implement CRUD for Employee Resource using Hibernate

### PR DESCRIPTION
**Pull request by TaulantFazliu**

### CODE REVIEW

File: src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java [hunk 1]
```diff
@@ -1,77 +1,54 @@
...
-import java.sql.ResultSet;
-import java.sql.SQLException;
-import java.time.LocalDate;
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
 import java.util.List;
 import java.util.Optional;
 
 @Repository
 public class JdbcEmployeeRepository implements EmployeeRepository {
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java:18]: You should create a new Repository impl named smth like JpaEmployeeRepository and then simply remove the Repository annotation from here and move it to the new class

Also you're going to see the benefit of IoC and dependency injection

File: src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java [hunk 2]
```diff
@@ -1,77 +1,54 @@
...
-                "on conflict (emp_no) do update set birth_date = :birthDate, first_name= :firstName, last_name= :lastName, gender= cast(:gender as gender), hire_date= :hireDate";
-        MapSqlParameterSource paramSource = new MapSqlParameterSource("empNo", employee.getEmpNo())
-                .addValue("birthDate", employee.getBirthDate())
-                .addValue("firstName", employee.getFirstName())
-                .addValue("lastName", employee.getLastName())
-                .addValue("gender", employee.getGender().toString())
-                .addValue("hireDate", employee.getHireDate());
-        namedParameterJdbcTemplate.update(sql, paramSource);
+
+        entityManager.persist(employee);
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java]: Always prefix with "this." when you access instance variables, always

File: src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java [hunk 3]
```diff
@@ -1,77 +1,54 @@
...
-        return employees;
+        return entityManager.createQuery("select e from Employee e", Employee.class).getResultList();
+
     }
 
     @Override
     public Optional<Employee> findById(Long emptNo) {
-        String sql = ("select * from employees where emp_no= :empNo");
-        MapSqlParameterSource paramSource = new MapSqlParameterSource("empNo", emptNo);
+        entityManager.find(Employee.class, emptNo);
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java]: One call here

File: src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java [hunk 4]
```diff
@@ -1,77 +1,54 @@
...
+
         try {
-            return Optional.ofNullable(namedParameterJdbcTemplate.queryForObject(sql, paramSource, new JdbcEmployeeRepository.EmployeeRowMapper()));
+            return Optional.ofNullable(entityManager.find(Employee.class, emptNo));
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java]: The second one here!!

Did you test for case when employee is not found? What exception did you get?

File: src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java [hunk 5]
```diff
@@ -1,77 +1,54 @@
...
             return Optional.empty();
         }
     }
 
     @Override
+    @Transactional
     public void deleteById(Long empNo) {
-        MapSqlParameterSource paramSource = new MapSqlParameterSource("empNo", empNo);
-        namedParameterJdbcTemplate.update("delete from employees where emp_no= :empNo", paramSource);
+        Employee employee = entityManager.find(Employee.class, empNo);
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java]: this. everywhere

File: src/main/resources/application.properties [hunk 6]
```diff
@@ -2,3 +2,10 @@ spring.datasource.url=jdbc:postgresql://localhost:5432/company_schema
 spring.datasource.username=postgres
 spring.datasource.password=password
 
+logging.level.org.hibernate.SQL=Debug
+spring.jpa.show-sql=true
+spring.jpa.properties.hibernate.format_sql=true
+
+spring.jpa.hibernate.ddl=none
+spring.jpa.hibernate.ddl-auto=create-drop
```
Comment (durimkryeziu) [inline src/main/resources/application.properties]: Keep in mind, schema creation is done by Flyway, Hibernate does it only on courses 😄

Make sure which one is used and whichever it is, set its value to none or validate is good as well

**Review by durimkryeziu — CHANGES_REQUESTED** [review]

File: src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java (hunk 4 above)
Comment (TaulantFazliu) [reply src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java]: EmployeeException and error 404 with message "Employee not found for id: {id}

File: src/main/resources/db/migration/V2__Add_employee_table.sql [hunk 7]
```diff
@@ -1,11 +1,9 @@
-CREATE TYPE gender AS ENUM ('M', 'F');
-
-CREATE TABLE employees
+CREATE TABLE employee
```
Comment (durimkryeziu) [inline src/main/resources/db/migration/V2__Add_employee_table.sql]: let's keep table names in plural, so `employees`

File: src/main/resources/db/migration/V2__Add_employee_table.sql [hunk 8]
```diff
@@ -1,11 +1,9 @@
...
 (
     emp_no     INTEGER     NOT NULL PRIMARY KEY,
     birth_date DATE        NOT NULL,
     first_name VARCHAR(14) NOT NULL,
     last_name  VARCHAR(16) NOT NULL,
-    gender     GENDER      NOT NULL,
+    gender     VARCHAR(2)  NOT NULL,
```
Comment (durimkryeziu) [inline src/main/resources/db/migration/V2__Add_employee_table.sql]:
```suggestion
    gender     VARCHAR(1)  NOT NULL,
```

Maybe?

File: src/main/java/com/employeesService/EmployeesService/model/Employee.java [hunk 9]
```diff
@@ -1,15 +1,21 @@
 package com.employeesService.EmployeesService.model;
 
+
+import javax.persistence.*;
 import java.time.LocalDate;
 
+@Entity(name = "employee")
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/model/Employee.java]:
```suggestion
@Entity(name = "employees")
```

File: src/main/java/com/employeesService/EmployeesService/model/Employee.java [hunk 10]
```diff
@@ -1,15 +1,21 @@
...
 
+    @Id
     private Long empNo;
     private LocalDate birthDate;
     private String firstName;
     private String lastName;
+    @Enumerated(EnumType.STRING)
+    @Column(name = "gender")
     private Gender gender;
     private LocalDate hireDate;
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/model/Employee.java]: Annotate each property with `@Column(name = "its db column name")`

File: src/main/java/com/employeesService/EmployeesService/model/Employee.java [hunk 11]
```diff
@@ -60,4 +66,27 @@ public LocalDate getHireDate() {
...
+//
+//    @Override
+//    public Employee.Gender convertToEntityAttribute(Character to) {
+//        Employee.Gender g = Employee.Gender.M;
+//        if ('f' == to){
+//            g = Employee.Gender.F;
+//    }
+//        return g;
+//}
+//}
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/model/Employee.java]: Commented code!!! Delete it

File: src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java [hunk 12]
```diff
@@ -1,77 +1,49 @@
...
-            employee.setEmpNo(rs.getLong("emp_no"));
-            employee.setBirthDate(rs.getObject("birth_date", LocalDate.class));
-            employee.setFirstName(rs.getString("first_name"));
-            employee.setLastName(rs.getString("last_name"));
-            employee.setGender(Employee.Gender.valueOf(rs.getString("gender")));
-            employee.setHireDate(rs.getObject("hire_date", LocalDate.class));
-            return employee;
-        }
-    }
 }
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/repository/JdbcEmployeeRepository.java]: Revert all this how it was implemented before with Jdbc, as the repository name is `Jdbc...`

File: src/main/java/com/employeesService/EmployeesService/repository/JpaEmployeeRepository.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+public class JpaEmployeeRepository implements EmployeeRepository {
+
+    @PersistenceContext
+    private EntityManager entityManager;
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/repository/JpaEmployeeRepository.java]: Didn't constructor injection work here?

File: src/main/java/com/employeesService/EmployeesService/repository/JpaEmployeeRepository.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Override
+    @Transactional
+    public Employee save(Employee employee) {
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/repository/JpaEmployeeRepository.java:26]: Did you verify update case is working as expected?

File: src/main/java/com/employeesService/EmployeesService/repository/JpaEmployeeRepository.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    @Override
+    public List<Employee> findAll() {
+        return this.entityManager.createQuery("select e from employee e", Employee.class).getResultList();
+
+    }
+
+    @Override
+    public Optional<Employee> findById(Long emptNo) {
+        try {
+            return Optional.ofNullable(this.entityManager.find(Employee.class, emptNo));
```
Comment (durimkryeziu) [inline src/main/java/com/employeesService/EmployeesService/repository/JpaEmployeeRepository.java:41]: Did you try the case when empNo is not found? What exception is thrown? Is your catch block catching it?

**Review by durimkryeziu — CHANGES_REQUESTED** [review]

**Review by durimkryeziu — APPROVED** [review]

**Pull request merged**